h/api/parse_document_claims.py

```diff
@@ -0,0 +1,215 @@
+
+    return document_uris
+
+
+def document_metas_from_data(document_data, claimant):
+    """
+    Return a list of document meta dicts for the given document data.
+
+    Returns one document meta dict for each document metadata claim in
+    document_data.
+
+    Each dict can be used to init a DocumentMeta object directly:
+
+        document_meta = DocumentMeta(**document_meta_dict)
+
+    :param document_data: the "document" sub-object that the client POSTed to
+        the API as part of a new or updated annotation
+    :type document_data: dict
+
+    :param claimant: the URI that the browser was at when this annotation was
+        created (the top-level "uri" field of the annotation)
+    :type claimant: unicode
+
+    :returns: a list of zero or more document meta dicts
+    :rtype: list of dicts
+
+    """
+    def transform_meta_(document_meta_dicts, items, path_prefix=None):
+        """Fill document_meta_dicts with document meta dicts for the items."""
+        if path_prefix is None:
+            path_prefix = []
+
+        for key, value in items.iteritems():
+            keypath = path_prefix[:]
+            keypath.append(key)
+
+            if isinstance(value, dict):
+                transform_meta_(document_meta_dicts,
+                                value,
+                                path_prefix=keypath)
+            else:
+                if not isinstance(value, list):
+                    value = [value]
+
+                document_meta_dicts.append({
+                    'type': '.'.join(keypath),
+                    'value': value,
+                    'claimant': claimant,
+                    'claimant_normalized':
+                        uri.normalize(claimant).decode('utf-8'),
+                })
+
+    items = {k: v for k, v in document_data.iteritems() if k != 'link'}
+    document_meta_dicts = []
+    transform_meta_(document_meta_dicts, items)
+    return document_meta_dicts
+
+
+def document_uris_from_links(link_dicts, claimant):
+    """
+    Return document URI dicts for the given document.link data.
+
+    Process a document.link list of dicts that the client submitted as part of
+    an annotation create or update request and return document URI dicts for
+    all of the document equivalence claims that it makes.
+
+    """
+    document_uris = []
+    for link in link_dicts:
+
+        # Disregard self-claim URLs as they're added separately later.
+        if link.keys() == ['href'] and link['href'] == claimant:
+            continue
+
+        # Disregard doi links as these are being added separately from the
+        # highwire and dc metadata later on.
+        if link.keys() == ['href'] and link['href'].startswith('doi:'):
+            continue
+
+        uri_ = link['href']
+        type_ = None
+
+        # Handle Highwire PDF links.
+        if set(link.keys()) == set(['href', 'type']):
+            if link['type'] == 'application/pdf':
+                type_ = 'highwire-pdf'
```

Comment:
I was thinking about this case recently. Instead of relying on the annotator doing the right thing ([this](https://github.com/openannotation/annotator/blob/v1.2.x/src/plugin/document.coffee#L122-L128)), wouldn't it be safer to get the `highwire-pdf` DocumentURI from the highwire metadata, similar to what we do for the doi uris?